bluetooth: host: conn: Utilize conn_set_state in conn_destroy

sean-madigan opened the pull request:
fixes https://github.com/zephyrproject-rtos/zephyr/issues/54542 , which highlighted a bug in bt_disable, where if you have pending tx packets for a connection and then call bt_disable, it will assert. 

Previously we would call conn_cleanup for all connections,
however this had multiple problems such as:
-Not destroying conn->tx_pending packets and causing an assert
-Calling conn_cleanup for connectable adv, triggering
disconnected callback

Now we will use bt_conn_set_state to manage the teardown of
connections correctly. First, if in connected or disconnecting,
set state to disconnect complete. Then for all states, set
state to disconnected. This will carry out proper cleanup
and teardown when required.

To do this, it was required to keep the tx_thread open as
this is used in disconnecting teardown, so this has been
moved to below the conn_cleanup_all call.

Signed-off-by: Sean Madigan <sean.madigan@nordicsemi.no>

Thalley commented on subsys/bluetooth/host/conn.c:
```diff
@@ -750,8 +750,14 @@ static void conn_destroy(struct bt_conn *conn, void *data)
 	k_work_cancel_delayable(&conn->deferred_work);
 
 	conn->pending_no_cb = 0;
+	conn->tx_pending.head = NULL;
+	conn->tx_pending.tail = NULL;
```
Alternatively we can call `sys_slist_init(conn->tx_pending)` which I guess does the same, but using the `slist` API

Thalley commented on subsys/bluetooth/host/conn.c:
```diff
@@ -750,8 +750,14 @@ static void conn_destroy(struct bt_conn *conn, void *data)
+	conn->tx_pending.tail = NULL;
 
-	conn_cleanup(conn);
+	if (conn->state == BT_CONN_CONNECTED ||
+		conn->state == BT_CONN_DISCONNECTING ||
+		conn->state == BT_CONN_DISCONNECT_COMPLETE) {
```
```suggestion
	if (conn->state == BT_CONN_CONNECTED ||
	    conn->state == BT_CONN_DISCONNECTING ||
	    conn->state == BT_CONN_DISCONNECT_COMPLETE) {
```

sean-madigan committed:
bluetooth: host: conn: Utilize conn_set_state in conn_destroy

Previously we would call conn_cleanup for all connections,
however this had multiple problems such as:
-Not destroying conn->tx_pending packets and causing an assert
-Calling conn_cleanup for connectable adv, triggering
disconnected callback

Now we will use bt_conn_set_state to manage the teardown of
connections correctly. First, if in connected or disconnecting,
set state to disconnect complete. Then for all states, set
state to disconnected. This will carry out proper cleanup
and teardown when required.

To do this, it was required to keep the tx_thread open as
this is used in disconnecting teardown, so this has been
moved to below the conn_cleanup_all call.

Signed-off-by: Sean Madigan <sean.madigan@nordicsemi.no>

Thalley commented:
@sean-madigan Have this been verified to fix the issue? 

If so, please add `fixes https://github.com/zephyrproject-rtos/zephyr/issues/54542` in the PR description

alwa-nordic reviewed (changes requested):
The references in `conn->tx_pending` must be properly destroyed using `conn_tx_destroy` here. All callbacks shall be invoked as usual; this is a promise to higher layers.

If we generate a 'disconnected' event without a corresponding previous 'connected' event, then there must be a bug in the monster `bt_conn_set_state`.

Thalley commented on subsys/bluetooth/host/conn.c:
```diff
@@ -747,11 +747,10 @@ static void conn_cleanup(struct bt_conn *conn)
-	conn->pending_no_cb = 0;
-
-	conn_cleanup(conn);
+	if (conn->state == BT_CONN_CONNECTED ||
+	    conn->state == BT_CONN_DISCONNECTING) {
```
we don't need to handle the `connecting` states?

sean-madigan replied:
As `bt_conn_set_state` has handling for the state change `BT_CONN_CONNECTING->BT_CONN_DISCONNECTED` I thought this was a safe operation.

Thalley replied:
Hmm, looking at `bt_conn_set_state`, is there actually any reason why we don't unconditionally call `bt_conn_set_state(BT_CONN_DISCONNECT_COMPLETE)`?

sean-madigan replied:
As there is specific handling for changing state `BT_CONN_CONNECTING/BT_CONN_CONNECTING_ADV` etc to `BT_CONN_DISONNECTED`, whereas `BT_CONN_CONNECTED/BT_CONN_DISCONNECTING` are designed to go through `BT_CONN_DISCONNECT_COMPLETE`
